frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -870,6 +967,618 @@ pub mod pallet {
+                Self::deposit_event(Event::<T>::ClaimRewardsAndDepositToBeneficiary(
+                    original_staker.clone(),
+                    contract_id.clone(),
+                    era.clone(),
+                    staker_reward,
+                    target,
+                ))
+            } else {
+                // if it's found we check if it's active and if it's the same smart contract
+                ensure!(
+                    beneficiary_info
+                        .clone()
+                        .expect("beneficiary is not found")
+                        .contract_id
+                        == SmartContractAddress::Address(contract_id.clone()),
+                    Error::<T>::InvalidBeneficiary
+                );
+
+                // ensure the beneficiary is active for receiving rewards
+                ensure!(
+                    beneficiary_info
+                        .clone()
+                        .expect("beneficiary is not found")
+                        .active,
+                    Error::<T>::BeneficiaryNotActive
+                );
+
+                ensure!(
+                    list_of_beneficiaries
+                        .iter()
+                        .find(|x| x.active == true)
+                        .is_some(),
+                    Error::<T>::BeneficiaryNotActive
+                );
+
+                // Check if we really do have the beneficiary
+                ensure!(
+                    list_of_beneficiaries
+                        .iter()
+                        .find(|x| x.account != target)
+                        .is_none(),
+                    Error::<T>::BeneficiaryNotFound
+                );
+
+                // update the existing beneficiary
+
+                // unwrap is safe because we checked if it's none before
+                let mut beneficiary = beneficiary_info.unwrap();
+                beneficiary.amount += staker_reward;
+
+                // transfer reward to the beneficiary
+                T::Currency::resolve_creating(&target, reward_imbalance);
+
+                RewardBeneficiaries::<T>::insert(&original_staker, &target, beneficiary);
+
+                Self::deposit_event(Event::<T>::ClaimRewardsAndDepositToBeneficiary(
+                    original_staker.clone(),
+                    contract_id.clone(),
+                    era.clone(),
+                    staker_reward.clone(),
+                    target,
+                ));
+            }
+
+            // set reward destination instead of relying on the default
+            let mut ledger = Self::ledger(&original_staker);
+            ledger.reward_destination = RewardDestination::BeneficiaryBalance;
+
+            // update ledger
+            Ledger::<T>::insert(&original_staker, ledger);
+            Self::update_staker_info(&original_staker, &contract_id, staker_info);
+            Ok(())
+        }
```

Comment:
Please note that this function has almost 200 lines and as that, it's very difficult to review :slightly_frowning_face: